src/query/api/v1/handler/namespace/update.go

```diff
@@ -0,0 +1,221 @@
+	if err != nil {
+		return nil, xhttp.NewParseError(err, http.StatusBadRequest)
+	}
+
+	updateReq := new(admin.NamespaceUpdateRequest)
+	if err := jsonpb.Unmarshal(bytes.NewReader(rBody), updateReq); err != nil {
+		return nil, xhttp.NewParseError(err, http.StatusBadRequest)
+	}
+
+	if err := validateUpdateRequest(updateReq); err != nil {
+		err := fmt.Errorf("unable to validate update request: %w", err)
+		return nil, xhttp.NewParseError(err, http.StatusBadRequest)
+	}
+
+	return updateReq, nil
+}
+
+// Ensure that only fields we allow to be updated (e.g. retention period) are
+// non-zero. Uses reflection to be resilient against adding more immutable
+// fields to namespaceOptions but forgetting to validate them here.
+func validateUpdateRequest(req *admin.NamespaceUpdateRequest) error {
+	if req.Name == "" {
+		return errEmptyNamespaceName
+	}
+
+	if req.Options == nil {
+		return errEmptyNamespaceOptions
+	}
+
+	if req.Options.RetentionOptions == nil {
+		return errEmptyRetentionOptions
+	}
+
+	optsVal := reflect.ValueOf(*req.Options)
+
+	for i := 0; i < optsVal.NumField(); i++ {
+		field := optsVal.Field(i)
+		fieldName := optsVal.Type().Field(i).Name
+		if !field.IsZero() && fieldName != fieldNameRetentionOptions {
+			return fmt.Errorf("%s: %w", fieldName, errNamespaceFieldImmutable)
+		}
+	}
+
+	optsVal = reflect.ValueOf(*req.Options.RetentionOptions)
+	for i := 0; i < optsVal.NumField(); i++ {
+		field := optsVal.Field(i)
+		fieldName := optsVal.Type().Field(i).Name
+		if !field.IsZero() && fieldName != fieldNameRetetionPeriod {
+			return fmt.Errorf("%s.%s: %w", fieldNameRetentionOptions, fieldName, errNamespaceFieldImmutable)
+		}
+	}
+
+	return nil
+}
+
+// Update updates a namespace.
+func (h *UpdateHandler) Update(
+	updateReq *admin.NamespaceUpdateRequest,
+	opts handleroptions.ServiceOptions,
+) (nsproto.Registry, error) {
+	var emptyReg = nsproto.Registry{}
+
+	kvOpts := kv.NewOverrideOptions().
+		SetEnvironment(opts.ServiceEnvironment).
+		SetZone(opts.ServiceZone)
+
+	store, err := h.client.Store(kvOpts)
+	if err != nil {
+		return emptyReg, err
+	}
+
+	currentMetadata, version, err := Metadata(store)
+	if err != nil {
+		return emptyReg, err
+	}
+
+	updateID := ident.StringID(updateReq.Name)
+	newMDs := make([]namespace.Metadata, 0, len(currentMetadata))
+	for _, ns := range currentMetadata {
+		// Don't modify any other namespaces.
+		if !ns.ID().Equal(updateID) {
+			newMDs = append(newMDs, ns)
+			continue
+		}
+
+		// Replace targeted namespace with modified retention.
+		if newNanos := updateReq.Options.RetentionOptions.RetentionPeriodNanos; newNanos != 0 {
+			dur := namespace.FromNanos(newNanos)
+			opts := ns.Options().SetRetentionOptions(
+				ns.Options().RetentionOptions().SetRetentionPeriod(dur),
+			)
+			newMD, err := namespace.NewMetadata(ns.ID(), opts)
+			if err != nil {
+				return emptyReg, fmt.Errorf("error constructing new metadata: %w", err)
+			}
+			newMDs = append(newMDs, newMD)
+		}
```

Comment:
Shouldn't there be a "} else { newMDs = append(newMDs, newMD) }"?

Otherwise if you update retention for a namespace that doesn't exist it'll basically remove it yeah (since will not add to newMDs and will think it's a valid request)?

Maybe we should add a test for this?